### ISSUE TITLE
Access to message fields name/values at the MavlinkMessage/MavlinkMessage2 level (like toString())

### ISSUE THREAD

**Issue by Enlade** [dronefleet/mavlink#9]
In my code I have a hierarchical list box that I use to allow the user to see the current field values for each message the program receives (whether the program processes the message or not). You can see this same type of structure in QGroundControl and in Mission Planner where they allow you to view the Mavlink messages as they come in (building the structure with each message or updating existing messages). There is a possibility that this heirarchy could be more than just 2 levels, but for my purposes I just list off the message name/ID and then under each of those I list off each field name/value (fitting each into a single string/line). So, for my purpose I just created a ArrayList where the first element of the list is the message name/ID and the subsiquent elements are the field name/values. There are likely better ways of doing this, particularly since some fields may have further sub-fields (like the enum fields).

The problem is that this library only accomodates a call to toString for some arbitrary MavlinkMessage. It would be nice if the MavlinkMessage would also give me the information in a more managable way for displaying it in such a hierarchical list box; maybe as XML or some such. For instance, I might want to add code similar to the following to each of the message classes (in this instance it is for Attribute):

```
public ArrayList<MAVLinkFieldPair> getMAVLinkFields() {	
	ArrayList<MAVLinkFieldPair> mlf = new ArrayList<MAVLinkFieldPair>();

	mlf.add(new MAVLinkFieldPair("MAVLINK_MSG_ID_ATTITUDE", String.valueOf(MAVLINK_MSG_ID_ATTITUDE)));
	mlf.add(new MAVLinkFieldPair("time_boot_ms", String.valueOf(time_boot_ms)));
	mlf.add(new MAVLinkFieldPair("roll", String.valueOf(roll)));
	mlf.add(new MAVLinkFieldPair("pitch", String.valueOf(pitch)));
	mlf.add(new MAVLinkFieldPair("yaw", String.valueOf(yaw)));
	mlf.add(new MAVLinkFieldPair("rollspeed", String.valueOf(rollspeed)));
	mlf.add(new MAVLinkFieldPair("pitchspeed", String.valueOf(pitchspeed)));
	mlf.add(new MAVLinkFieldPair("yawspeed", String.valueOf(yawspeed)));
	    
	return(mlf);	
}
```

As it is now there is no such methods, so the only way to get to this information (without casting the message, etc) is to parse the result of the toString call.  For instance, I could use a function like this to give me similar results:

```
public void updateMAVLinkMessage(int pID, String pST) {
	ArrayList<MAVLinkFieldPair> mlf = new ArrayList<MAVLinkFieldPair>();

	int fPos1 = pST.indexOf('{');
	int fPos2 = pST.lastIndexOf('}');

	mlf.add(new MAVLinkFieldPair(pST.substring(0, fPos1), String.valueOf(pID)));
	
	String[] fields = pST.substring(fPos1+1, fPos2).split(",(?![^\\{]*[\\}])");
	
	for (int i=0; i<fields.length; i++) {
		fPos1 = fields[i].indexOf('{');
		if (fPos1<0) {
			fPos2 = fields[i].indexOf('=');
			mlf.add(new MAVLinkFieldPair(fields[i].substring(0, fPos2).trim(), String.valueOf(fields[i].substring(fPos2+1).trim())));								
		} else {
			fPos2 = fields[i].indexOf('=');
			String tempST1 = fields[i].substring(0, fPos2).trim();
			String tempST2 = "[";

			fPos2 = fields[i].lastIndexOf('}');
			String[] enumf = fields[i].substring(fPos1+1, fPos2).split(",");
			for (int j=0; j<enumf.length; j++) {
				if (j>0) tempST2 = tempST2 + ", ";

				fPos2 = enumf[j].indexOf('=');
				tempST2 = tempST2 + String.valueOf(enumf[j].substring(fPos2+1).trim());
			}
			tempST2 = tempST2 + "]";
			mlf.add(new MAVLinkFieldPair(tempST1, tempST2));							
		}
	}

	updateMAVLinkMessage(mlf);
}
```

Of course this is a bit messy and something that would need to be managed whenever there is any updates to the library (that might effect the toString format).  I am not even sure that I have covered the format of all messages (only ones I have run into so far).  So, it might be a good idea to have some way to get the message field values without having to cast the message (like is done with toString).

**Comment by benbarkay** [comment]
You can access field information by using the annotations. For instance:

```java
@Test
public void test() throws InterruptedException {
    CommandLong message = CommandLong.builder()
            .command(MavCmd.MAV_CMD_COMPONENT_ARM_DISARM)
            .param1(1)
            .param2(2989.0f)
            .targetSystem(1)
            .targetComponent(0)
            .build();

    Arrays.stream(message.getClass().getDeclaredMethods())
            .filter(f -> f.isAnnotationPresent(MavlinkFieldInfo.class))
            .forEach(f -> {
                try {
                    System.out.printf("%s = %s\n", f.getName(), f.invoke(message).toString());
                } catch (InvocationTargetException | IllegalAccessException e) {
                    e.printStackTrace();
                }
            });
}
```

**Comment by Enlade** [comment]
Yea, I thought of that too, but I was just trying to make a different point and didn't want to work so hard on a workaround example as to trace through annotations; so I just tossed in that little parser.  Still, since you wrote out the code I'll surely use it instead of my little parser (it is much better).

What I was trying to say was that it isn't exactly straight forward on how to do these things, and maybe if you added this method to your MavlinkMessage class, but returned the values in some format rather then outputting it to the console then that might be more easily useful to users (and more self evident).  At least they would see some method called, say getFields(), or some such, and they might naturally conclude its usefulness.  I mean, most of your users are likely to not have even used annotations or many of these other undocumented workarounds that you are proposing.  It might be helpful to make your library more easily usable to toss in some of these methods/classes/etc, just to make it easier for your users.  For me, this is fine.  I will use your function and be happy.

**Comment by benbarkay** [comment]
Like I've mentioned in another issue that you've opened, these are non-issues for people who come from a Java background, and are familiar with annotations, which is why I don't see it as a defect.

This goalpost can be pushed as far as we want it to, programming is not easy and there is a spectrum of proficiency. This library assumes you're at least familiar or are interested in getting familiar with the nuances of the language.

**Comment by Enlade** [comment]
Well, you should have an obligation to write library code for your target audience, and that is not CS majors with years of JAVA development behind them.  Like it or not, most of your audience is just hobbiests.  Still, I don't want to argue with you about good programming practices (like having good documentation and such), so I'll just agree to disagree with you.

**Comment by Enlade** [comment]
Mind you, I do want you to know, now that I had to work through all of your code, that the way you are doing your backend coding is very nice.  I want to make sure you understand that I am grateful for your efforts and respect your abilities as a JAVA programmer.  I'm just suggesting that the front end of your library can be much simpler without really changing any of your nice back end code.  I just want to be a user of your library.  I'm expecting it to be simple, not because I can't understand the complexity, but because I should not have to.

At this point I can read through your code and understand its complexity and create my own methods to utilize it to do what I need (in most cases), but why do you want me to do that?  Why not just add the methods in a way that makes it very easy for me to see what I need to do (even without documentation)?

So, please don't think of what I am saying in these threads as some attack on your abilities as a programmer. I know you are a very good programmer. All of my suggestions/questions really have little to do with modifying the architecture/design of your back end code, its mostly just adding some additional methods to the front end so that users can more easily get to what they are after.

Think of my suggestions/questions more as a documented experience of a user that is trying to use your library (without documentation).  What do you think a user is expecting when they approach your library?  Well, that is really what you should give them as a front end, even if you choose to do the backend in some creative, sophisticated, way.  Your user doesn't need to know any of that.  I mean, I do know it now, but I had to look behind the curtain and work through your code to get there, and as a programmer I feel a little dirty having to do it because I'm not supposed to do that when I use libraries.  Why do that to your users, particularly if they don't have multiple CS degrees and years of JAVA programming experience?

**Comment by Enlade** [comment]
Fair enough.  I'll keep posting, you take what you find useful, and maybe others will find it useful too.  I've got a project I'm working on as well and I have created a similar library for my project supporting MAVLink 1.0; but, I didn't want to continue to manage that library and I didn't want to add MAVLink 2.0 support; so, I am very grateful to not have to manage such a library.  I just want to focus on my application and I am thankful that you are helping me do that with your library.
